src/proxy.rs

```diff
@@ -1,8 +1,137 @@
 use std::fmt;
 use std::sync::Arc;
+use std::net::{SocketAddr, ToSocketAddrs};
 
 use hyper::client::connect::Destination;
-use {into_url, IntoUrl, Url};
+use {IntoUrl, Url};
+
+
+// The kinds of authentication supported to a proxy
+#[derive(Clone, Debug)]
+enum ProxyAuth {
+    Basic {
+        username: String,
+        password: String,
+    }
+}
+
+// The various proxy schemes we understand
+#[derive(Clone, Debug)]
+enum ProxySchemeKind {
+    Http,
+    Https,
+    Socks5,
+    Socks5h,
+}
+
+/// A particular scheme used for proxying requests.
+/// 
+/// For example, HTTP vs SOCKS5
+#[derive(Clone, Debug)]
+pub struct ProxyScheme {
+    kind: ProxySchemeKind,
+    uri: Option<::hyper::Uri>,
+    socket_addrs: Option<Vec<SocketAddr>>,
+    auth: Option<ProxyAuth>,
+}
+
+impl ProxyScheme {
+    /// Proxy traffic via the specified URL over HTTP
+    pub fn http<T: IntoUrl>(url: T) -> ::Result<Self> {
+        Ok(ProxyScheme {
+            kind: ProxySchemeKind::Http,
+            uri: Some(::into_url::to_uri(&try_!(url.into_url()))),
+            socket_addrs: None,
+            auth: None,
+        })
+    }
+
+    /// Proxy traffic via the specified URL over HTTPS
+    pub fn https<T: IntoUrl>(url: T) -> ::Result<Self> {
+        Ok(ProxyScheme {
+            kind: ProxySchemeKind::Https,
+            uri: Some(::into_url::to_uri(&try_!(url.into_url()))),
+            socket_addrs: None,
+            auth: None,
+        })
+    }
+
+    /// Proxy traffic via the specified socket address over SOCKS5
+    pub fn socks5<T: ToSocketAddrs>(socket_addr: T) -> ::Result<Self> {
```

Comment:
`ToSocketAddrs` can result in network calls (DNS lookups), is that actually desired? They'd even be blocking calls...

Reply:
Ideally there would be an asynchronous option, but I'm not sure the primitives exist yet for that? I do think this should happen on construction of the proxy scheme though, at least that way it can be controlled by the user (eg. they could put it on a thread pool), and doesn't need to happen multiple times if constructing multiple clients.

I'm open to suggestions though!

Reply:
The other `Proxy` types save a `Uri`, and thus can use new DNS entries as the program goes on, whereas in this case, the socket addresses are only looked up once. If following what curl does, `http_proxy="socks5h://[<username>:<password>@]<proxy_hostname>[:<proxy_port>]"` would mean we could store a `Uri` and do DNS the same as for HTTP.

Reply:
Hm, that makes sense, but I still don't think it should store a `Uri` - it should store a [`TargetAddr`-like thing](https://docs.rs/socks/0.3.1/socks/enum.TargetAddr.html) instead of a `SocketAddr`.

Forcing users to specify a proxy using the curl syntax is a PITA, because username and password must be URL-encoded to be valid `Uri`s (in fact I ran into this exact issue: curl has special parsing code that accepts invalid `Uri`s, whereas the rust `Uri` crate is much stricter and requires everything to be URL-encoded).

Reply:
True, I wouldn't recommend that people include the username and password in the URI, since that is deprecated and generally Bad. But the rest of it can be:

```rust
// intercepts all `http://` requests and sends over socks5h
let proxy = Proxy::http("socks5h://my.socks.local:1234")?
    .with_basic_auth("username", "password");
```

Reply:
It's not clear to me how that would interact with custom proxies

Reply:
I don't quite get what you're referring to. What custom proxies?

Reply:
`Proxy::custom`

Reply:
Oh OK. It seems you'd be able to return a `ProxyScheme::http(url).with_basic_auth()` inside `custom`, no?

Reply:
Right, but your example was:
```rust
let proxy = Proxy::http("socks5h://my.socks.local:1234")?
    .with_basic_auth("username", "password");
```

And not:
```rust
let proxy = Proxy::http(ProxyScheme::parse("socks5h://my.socks.local:1234")?
    .with_basic_auth("username", "password"))?;
```

Given that `Proxy` is an enum, it's a bit tricky to implement methods like `with_basic_auth` in a consistent way.